File: Trees/inorder_stack.cpp
/* Program for the Inorder Traversal without recursion and using stack*/
#include <iostream>
using namespace std;
#include <stack>
#include<bits/stdc++.h>
struct Node{
    int data;
    Node* left;
    Node* right;

    Node(int data){
        this->data =data;
        left = NULL;
        right = NULL; 
    }
};

void inorder(struct Node* root){
    stack <Node*> s;
    Node *curr = root;
    while(curr!=NULL || s.empty()==false){
        while(curr !=NULL){
            s.push(curr);
            curr = curr->left;
        }
        curr = s.top();
        s.pop();
        cout << curr->data << " ";
        curr = curr->right;

    }
}
int main() 
{ 
  
    /* Constructed binary tree is 
              1 
            /   \ 
          2      3 
        /  \ 
      4     5 
    */
    struct Node *root = new Node(1); 
    root->left        = new Node(2); 
    root->right       = new Node(3); 
    root->left->left  = new Node(4); 
    root->left->right = new Node(5); 
  
    inorder(root); 
    return 0; 
} 

File: Trees/inorder_stack.py
'''
Program for the Inorder traversal of a binary tree
without recursion using stack(list)
'''
class Node:
    def __init__(Self,key):
        self.key = key
        self.left = None
        self.right = None


def inorder(root):
    current = root
    s = []
    while True:
        if current is not None:
            s.append(current)
            current = current.left
        elif s:
            current = s.pop()
            print(current.key,end = ' ')
            current = current.right
        else:
            break

""" Constructed binary tree is 
            5
          /   \ 
         6     3             
       /  \    /
      4    5  2   """
  
root = Node(5) 
root.left = Node(2) 
root.right = Node(3)
root.left = Node(2) 
root.left.left = Node(4) 
root.left.right = Node(5) 

inorder(root)